pkg/oc/admin/diagnostics/diagnostics/cluster/app_create/main.go

```diff
@@ -0,0 +1,359 @@
+		{"http-retries", "Number of times to retry an HTTP request to the app", &d.httpRetries, AppCreateHttpRetriesDefault},
+		{"node-selector", "Node selector for where the test app should land", &d.nodeSelector, ""},
+		{"keep-app", "Do not delete the test app when complete", &d.keepApp, false},
+		{"result-dir", "Directory in which to write result details if desired", &d.writeResultDir, ""},
+	}
+}
+
+func getDefaultAppImage() string {
+	template := variable.NewDefaultImageTemplate()
+	return template.ExpandOrDie("deployer")
+}
+
+func (d *AppCreate) Complete(logger *log.Logger) error {
+	// project management
+	d.keepProject = d.keepProject || d.keepApp // don't delete project if keeping app
+	if d.project == "" && d.projectBase == "" {
+		return fmt.Errorf("%s project name cannot be empty", AppCreateName)
+	}
+	if d.project == "" {
+		// generate a project if not specified
+		d.project = names.SimpleNameGenerator.GenerateName(d.projectBase)
+	} else {
+		// when an existing project is specified, deleting it is likely to surprise the user, so don't
+		d.keepProject = true
+	}
+	if errs := kvalidation.IsDNS1123Label(d.project); len(errs) > 0 {
+		return fmt.Errorf("invalid project name '%s' for AppCreate: %v", d.project, errs)
+	}
+	// TODO: also test that route is valid under DNS952
+
+	// app management
+	if d.appName == "" {
+		return fmt.Errorf("%s app name cannot be empty", AppCreateName)
+	}
+	if errs := kvalidation.IsDNS1123Label(d.appName); len(errs) > 0 {
+		return fmt.Errorf("invalid app name '%s' for AppCreate: %v", d.appName, errs)
+	}
+	if err := kvalidation.IsValidPortNum(d.appPort); err != nil {
+		return fmt.Errorf("invalid app port %d for AppCreate: %v", d.appPort, err)
+	}
+	d.label = map[string]string{"app": d.appName}
+	d.labelSelector = fmt.Sprintf("app=%s", d.appName)
+
+	d.skipRouteConnect = d.skipRouteConnect || d.skipRouteTest // don't try to connect to route if skipping route test
+
+	return nil
+}
+
+func (d *AppCreate) CanRun() (bool, error) {
+	if d.SARClient == nil || d.AppsClient == nil || d.KubeClient == nil || d.ProjectClient == nil || d.RoleBindingClient == nil {
+		return false, fmt.Errorf("missing at least one client")
+	}
+	if d.PreventModification {
+		return false, fmt.Errorf("requires modifications: create a project and application")
+	}
+	return util.UserCan(d.SARClient, &authorization.ResourceAttributes{
+		Verb:     "create",
+		Group:    kapi.GroupName,
+		Resource: "namespace",
+	})
+}
+
+func (d *AppCreate) Check() types.DiagnosticResult {
+	d.out = types.NewDiagnosticResult(AppCreateName)
+	done := make(chan bool, 1)
+
+	// Jump straight to clean up if there is an interrupt/terminate signal while running diagnostic
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-sig
+		d.out.Warn("DCluAC001", nil, "Received interrupt; aborting diagnostic")
```

Comment:
does this actually abort the other gofunc?

Reply:
No, it doesn't. There is no way to abort a goroutine - it's collaborative concurrency, the goroutine has to want to stop. The main one keeps running after an interrupt, we're just not paying attention to it any longer. The only thing I could think of is to set up another channel and check it at various points to see if an interrupt occurred, but that seemed even messier. Do you have a different suggestion?

This is not new at all BTW, just code moved around.

Reply:
missed it was a move

Reply:
Most of the second commit is stuff moving around